src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs

```
@@ -1015,5 +1165,75 @@ private void AddSample<T>(List<T> samples, T sample)
+
+        private sealed class PendingItem
+        {
+            private readonly Action<InboundGrpcEvent> _callback;
+            private readonly Action<Exception> _faultHandler;
+            private CancellationTokenRegistration _ctr;
+            private int _state;
+
+            public PendingItem(Action<InboundGrpcEvent> callback, Action<Exception> faultHandler)
+            {
+                _callback = callback;
+                _faultHandler = faultHandler;
+            }
+
+            public PendingItem(Action<InboundGrpcEvent> callback, Action<Exception> faultHandler, TimeSpan timeout)
+                : this(callback, faultHandler)
+            {
+                var cts = new CancellationTokenSource();
+                cts.CancelAfter(timeout);
+                _ctr = cts.Token.Register(static state => ((PendingItem)state).OnTimeout(), this);
+            }
+
+            public bool IsComplete => Volatile.Read(ref _state) != 0;
+
+            private bool MakeComplete() => Interlocked.CompareExchange(ref _state, 1, 0) == 0;
+
+            public void SetResult(InboundGrpcEvent message)
+            {
+                _ctr.Dispose();
+                _ctr = default;
+                if (MakeComplete() && _callback != null)
+                {
+                    try
+                    {
+                        _callback.Invoke(message);
+                    }
+                    catch (Exception fault)
+                    {
+                        try
+                        {
+                            _faultHandler?.Invoke(fault);
+                        }
+                        catch
+                        {
+                        }
+                    }
+                }
+            }
+
+            private void OnTimeout()
+            {
+                if (MakeComplete() && _faultHandler != null)
+                {
+                    try
+                    {
+                        throw new TimeoutException();
```

Do we need to throw an new exception just to catch it and handle? Could we just do `_faultHandler(timeout);` in the first try?